File: 2021/day-2/javascript/solve-puzzle-2.mjs
import { readFileSync } from 'node:fs';

const input = readFileSync('../input.txt', 'utf-8').split('\n');

let horizontalPosition = 0;
let depth = 0;
let aim = 0;

for (const instruction of input) {
  const instructionLine = instruction.split(' ');
  const arg = Number(instructionLine[1]);

  switch (instructionLine[0]) {
    case 'forward':
      horizontalPosition += arg;
      depth += aim * arg;
      break;
    case 'up':
      aim -= arg;
      break;
    case 'down':
      aim += arg;
  }
}

console.log(horizontalPosition * depth);

File: 2021/day-2/python/solve-puzzle-2.py
with open("../input.txt", "r") as handle:
    horizontal_position = 0
    depth = 0
    aim = 0

    for instruction in handle.read().splitlines():
        instruction_line = instruction.split(" ")
        arg = int(instruction_line[1])
        instruction_line = instruction_line[0]

        if instruction_line == "forward":
            horizontal_position += arg
            depth += aim * arg
        elif instruction_line == "up":
            aim -= arg
        elif instruction_line == "down":
            aim += arg

    print(horizontal_position * depth)
